File: 2022/17/javascript/main.mjs
import fs from "node:fs";
import { SHAPES, createChamber } from "./chamber.mjs";

// --- Constants ---
const INPUT_FILE = "../input/input.txt";
const LIMIT_1 = 2022;
const LIMIT_2 = 1_000_000_000_000;

// --- Runs the simulations ---
function sim(flows, limit) {
	let count = 0;
	let rocks = 1;
	const gcd = SHAPES.length * flows.length; // Greates Common Divisor
	let chamber = createChamber();
	chamber.addShape(SHAPES[count % SHAPES.length]);

	// --- Variables for tracking a pattern ---
	let prevHeight = 0;
	let prevRocks = 0;
	let deltaHeight = 0;
	let deltaRocks = 0;
	let simulated = 0;

	// --- Iterate until the limit is reached ---
	while (true) {
		if (count > 0 && count % gcd === 0) {
			const height = chamber.getHeight();
			const dh = height - prevHeight;
			const dr = rocks - prevRocks;
			if (dh === deltaHeight && dr === deltaRocks) {
				const rate = Math.floor((limit - rocks) / dr);
				simulated = dh * rate + 1;
				rocks = limit - ((limit - rocks) % dr);
			}
			prevHeight = height;
			prevRocks = rocks;
			deltaHeight = dh;
			deltaRocks = dr;
		}

		// --- Normal simulation ---
		chamber.move(flows[count % flows.length]);
		if (!chamber.down()) {
			if (rocks > limit) {
				return chamber.getHeight() + simulated;
			}
			chamber.addShape(SHAPES[rocks % SHAPES.length]);
			rocks++;
		}
		count++;
	}
}

// --- Reads and parse the input file ---
const text = fs.readFileSync(INPUT_FILE, "utf8");
const flows = Array.from(text.trim());

// --- Puzzle 1 ---
let height = sim(flows, LIMIT_1);
console.log("1. Tower height:", height.toLocaleString());

// --- Puzzle 2 ---
height = sim(flows, LIMIT_2);
console.log("2. Tower height:", height.toLocaleString());

File: 2022/17/c/main.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <file_reader.h>
#include <lib.h>

// --- Constants ---
#define INPUT_FILE "../input/input.txt"
#define LIMIT_1 2022L
#define LIMIT_2 1000000000000L

// Runs the simulations
long sim(char* flows, long limit) {
	int flows_len = strlen(flows);
	long count = 0;
	long rocks = 1;
	long gcd = SHAPE_LEN * flows_len;
	Chamber chamber = { .root = row_create_node(), .shape_len = 0 };
	chamber_add_shape(&chamber, 0);

	// --- Variables for tracking a pattern ---
	long prev_height = 0;
	long prev_rocks = 0;
	long delta_height = 0;
	long delta_rocks = 0;
	long simulated = 0;
	long height, dh, dr, rate;

	// --- Iterate until the limit is reached ---
	while (1) {
		// --- Tries to find a pattern ---
		if (count > 0 && count % gcd == 0) {
			height = chamber_height(&chamber);
			dh = height - prev_height;
			dr = rocks - prev_rocks;

			// --- Pattern found ---
			if (dh == delta_height && dr == delta_rocks) {
				rate = (limit - rocks) / dr;
				simulated = dh * rate + 1;
				rocks = limit - ((limit - rocks) % dr);
			}
			prev_height = height;
			prev_rocks = rocks;
			delta_height = dh;
			delta_rocks = dr;
		}

		// --- Normal simulation ---
		chamber_move(&chamber, flows[count % flows_len]);
		if (!chamber_move(&chamber, DOWN)) {
			if (rocks > limit) {
				height = chamber_height(&chamber) + simulated;
				chamber_free(&chamber);
				return height;
			}
			chamber_add_shape(&chamber, rocks % SHAPE_LEN);
			rocks++;
		}
		count++;
	}
}

// --- Main function ---
int main(void) {
	// --- Read the input file ---
	char* text = NULL;
	READ_FULL(text, INPUT_FILE);
	int len = strlen(text);
	if (text[len - 1] == '\n') text[len - 1] = 0;

	// --- Puzzle 1 ---
	long height = sim(text, LIMIT_1);
	printf("1. The tower is %ld units tall\n", height);

	// --- Puzzle 2 ---
	height = sim(text, LIMIT_2);
	printf("2. The tower is %ld units tall\n", height);

	// --- Free resources ---
	free(text); text = NULL;

	return EXIT_SUCCESS;
}
